### PR TITLE
[DDW-539, DDW-541] Fund3 voting registration

### PR DESCRIPTION
**Pull request by nikolaglumac**
This PR implements Catalyst "Voting registration" feature (https://github.com/input-output-hk/daedalus/pull/2204) into `develop` branch.
It also improves the syncing states of ITN redemption and Staking screens.

## Documentation
- Original test plan: https://input-output.atlassian.net/wiki/spaces/QA/pages/1824293038/Voting+Test+Plan
- New feature flow: https://docs.google.com/document/d/1SluIjNM5qF7lB7gp1_mvBRCZcJQG9BG-gwWMy38qlq0/edit#

## Designs
https://zpl.io/aMe0XwN

## Todos

- [x] Inspect QR code on the "Info" screen on dark themes (@nikolaglumac)
- [x] Implement "Save as PDF" feature (@daniloprates)
- [x] QR code step text alignment and warning text color (@nikolaglumac)
- [x] PIN entry component behaviour (@nikolaglumac)
- [x] Use the same no-wallets pattern as on the "Delegation" wizard
- [x] Implement animation for the progress bar on "Confirm" step
- [x] Implement checkboxes logic on the "QR code" step
- [x] Implement support for all Daedalus themes
- [x] Inject Japanese translations

### Known issues
- [x] ~Counter stops when screen is not active~ [VIT-739](https://jira.iohk.io/browse/VIT-739) **No longer relevant due to change in UI/UX logic (checking transaction confirmations vs 60 mins time period)**
- [x] Can't calculate voting registration transaction fees or submit the transaction in case you start the voting registration close to Daedalus launch (to be fixed in https://github.com/input-output-hk/cardano-wallet/pull/2468) - **Confirmed to be resolved by https://github.com/input-output-hk/daedalus/pull/2325** 

### Code quality improvements
- [x] Implementing fixes for the issues listed in the [Code Quality Review](https://github.com/input-output-hk/daedalus/pull/2315#pullrequestreview-575247743)
- [x] Improve container/store logic to avoid strange "jumping" between the steps in the dialog (e.g. when you enter wrong wallet passphrase)
- [x] Rename all components and containers from "voting-add" to "voting-registration"

### UI/UX improvements
- [x] Implementing fixes for the issues listed in the [UI/UX Implementation Review](https://docs.google.com/document/d/1MvSo4Y6tToVlNPMa9dnte1jnG8iEkMvSpeVDpEtzh9E/edit?usp=sharing)
- [x] Add loading state for the voting UI during Daedalus syncing process
- [x] Add [AppStore link for downloading Catalyst Voting App](https://apps.apple.com/in/app/catalyst-voting/id1517473397)
- [x] Disable voting UI until Daedalus is fully synced
- [x] Disable voting registration for hardware wallets (incl. a display of a dedicated error message)
- [x] Disable voting registration for Byron wallets (incl. a display of a dedicated error message)
- [x] Implement voting registration user flow changes (show on the chart below)
- [x] Implement text copy improvements listed in [Text Copy Review](https://docs.google.com/document/d/1SluIjNM5qF7lB7gp1_mvBRCZcJQG9BG-gwWMy38qlq0/edit?ts=60103568) (e.g. [the link on the Voting Wizard Step#2](https://docs.google.com/document/d/1_syHaBNjhFcB5qg_NJoBfIZIBy_ha2n042XdrDASYgY/edit#heading=h.u296e49etfth) leads to an empty Google Document)

## Voting registration flow

Screenshot 2021-01-25 at 17 47 01

### Staking screen during Daedalus syncing

Screenshot 2021-02-05 at 18 51 29
Screenshot 2021-02-05 at 18 51 23

### ITN rewards redemption screen during Daedalus syncing
 
Screenshot 2021-02-05 at 18 53 59
Screenshot 2021-02-05 at 18 54 06

### Voting screen during Daedalus syncing

Screenshot 2021-02-05 at 18 51 19
Screenshot 2021-02-05 at 18 51 32

### Voting screen when there are no Shelley-compatible wallets

Screenshot 2021-02-05 at 18 50 15
Screenshot 2021-02-05 at 18 50 22

### Voting info screen

Screenshot 2021-02-05 at 18 57 47
Screenshot 2021-02-05 at 18 57 57
Screenshot 2021-02-05 at 18 57 52
Screenshot 2021-02-05 at 18 58 00

### Voting registration wizard

Screenshot 2021-02-05 at 18 59 00
Screenshot 2021-02-05 at 19 01 39
Screenshot 2021-02-05 at 19 01 02
Screenshot 2021-02-05 at 19 01 25
Screenshot 2021-02-05 at 19 19 58
Screenshot 2021-02-05 at 19 20 08
Screenshot 2021-02-05 at 19 20 25
Screenshot 2021-02-05 at 19 20 12
Screenshot 2021-02-05 at 19 21 08
Screenshot 2021-02-05 at 19 21 26
Screenshot 2021-02-05 at 19 21 12
Screenshot 2021-02-05 at 19 21 28
Screenshot 2021-02-05 at 19 34 52
Screenshot 2021-02-05 at 19 34 33
Screenshot 2021-02-05 at 19 24 19
Screenshot 2021-02-05 at 19 40 21
Screenshot 2021-02-05 at 19 32 56
Screenshot 2021-02-05 at 19 47 34
Screenshot 2021-02-05 at 19 33 29
Screenshot 2021-02-05 at 19 47 36
Screenshot 2021-02-05 at 19 33 35
Screenshot 2021-02-05 at 19 47 41
Screenshot 2021-02-05 at 19 33 40
Screenshot 2021-02-05 at 19 47 44
Screenshot 2021-02-05 at 19 33 52
Screenshot 2021-02-05 at 19 47 47
Screenshot 2021-02-05 at 19 33 55
Screenshot 2021-02-05 at 19 47 49
Screenshot 2021-02-05 at 19 34 06
Screenshot 2021-02-05 at 19 47 54

### QR code PDF

[fund3-voting-registration-Shelley Wallet-2021-02-05T193357.0674.pdf](https://github.com/input-output-hk/daedalus/files/5934634/fund3-voting-registration-Shelley.Wallet-2021-02-05T193357.0674.pdf)
[fund3-有権者登録-Shelley Wallet-2021-02-05T194749.0956.pdf](https://github.com/input-output-hk/daedalus/files/5934636/fund3-.-Shelley.Wallet-2021-02-05T194749.0956.pdf)

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)
- [x] Run regression tests for existing Daedalus features
- [x] Run tests of the voting feature

---

## Testing Summary

- 1st regression: [Testing document link](https://docs.google.com/spreadsheets/d/1yBMY-zbhqkNNVsqbdyjYcTC-_u7MWiq4tNVHFAf0Q3c/edit?usp=sharing)
- 2nd regression: [Testing document link](https://docs.google.com/spreadsheets/d/1NaIOxAwzaQpQbWHYr18OmV_ZEmKhqmbJ7TfeolOFo-c/edit?usp=sharing)

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by nikolaglumac** [comment]
@gnpf @ManusMcCole @mioriohk here's the document with a list of UI/UX defects that @a-rukin prepared: https://docs.google.com/document/d/1MvSo4Y6tToVlNPMa9dnte1jnG8iEkMvSpeVDpEtzh9E/edit

Please coordinate with @a-rukin and add all the defects you might find to the same doc 🙏

File: source/renderer/app/api/utils/requestBinary.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// @flow
```
Comment (nikolaglumac) [inline source/renderer/app/api/utils/requestBinary.js:1]: Do we really need this `requestBinary`? We should rather use the existing `request` with modifications if they are needed.

File: source/renderer/app/api/voting/requests/createWalletSignature.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// @flow
+import type { RequestConfig } from '../../common/types';
+import type { SignatureParams } from '../types';
+import { request } from '../../utils/requestBinary';
```
Comment (nikolaglumac) [inline source/renderer/app/api/voting/requests/createWalletSignature.js:4]: Do we really need this `requestBinary`? We should rather use the existing `request` with modifications if they are needed.

File: source/renderer/app/stores/VotingStore.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    this.setIsVotingRegistrationTransactionPending(false);
+  };
+
+  generateVotingRegistrationKey = async () => {
+    const Modules = await wallet;
+    const key = Modules.Ed25519ExtendedPrivate;
+    this.setVotingRegistrationKey(key.generate());
+  };
+
+  getHexFromBech32 = async (key: string) => {
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:196]: Return type declaration is missing.

File: source/renderer/app/stores/TransactionsStore.js [hunk 4]
```diff
@@ -395,6 +395,10 @@ export default class TransactionsStore extends Store {
     return new Request(this.api.ada.getWithdrawals);
   };
 
+  _getTransaction = (): Request<WalletTransaction> => {
```
Comment (nikolaglumac) [inline source/renderer/app/stores/TransactionsStore.js:398]: This should be renamed to `_getTransactionRequest`.

File: source/renderer/app/stores/VotingStore.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+      PASSWORD,
+      this.votingRegistrationKey.bytes()
+    );
+    this.setQrCode(formattedArrayBufferToHexString(encrypt));
+  };
+
+  // Check voting registration transaction state and reset pending state when transaction is "in_ledger"
+  checkVotingRegistrationTransaction = async () => {
+    const transaction = await this.stores.transactions
+      ._getTransaction()
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:177]: This should be renamed to `_getTransactionRequest`.

File: source/renderer/app/stores/VotingStore.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    this.countdownRemaining = VOTING_COUNTDOWN_INTERVAL;
+    if (this.countdownTimerInterval) clearInterval(this.countdownTimerInterval);
+    this.countdownTimerInterval = setInterval(() => {
+      if (this.countdownRemaining > 0) {
+        action(() => this.countdownRemaining--)();
+      } else if (this.countdownTimerInterval != null) {
+        clearInterval(this.countdownTimerInterval);
+        this.checkVotingRegistrationTransaction();
+      }
+    }, 1000);
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:73]: 1000 should be extracted into a configuration variable.

File: source/renderer/app/stores/SidebarStore.js [hunk 7]
```diff
@@ -82,6 +82,7 @@ export default class SidebarStore extends Store {
       [categories.STAKING.name]: isShelleyActivated,
       [categories.REDEEM_ITN_REWARDS.name]: true,
       [categories.SETTINGS.name]: true,
+      [categories.VOTING.name]: true,
```
Comment (nikolaglumac) [inline source/renderer/app/stores/SidebarStore.js:85]: We need to decide on which networks Voting menu should be visible.

File: source/renderer/app/routes-config.js [hunk 8]
```diff
@@ -30,6 +30,9 @@ export const ROUTES = {
     SETTINGS: '/wallets/:id/settings',
     UTXO: '/wallets/:id/utxo',
   },
+  VOTING: {
+    ADD: '/voting-catalyst/add',
```
Comment (nikolaglumac) [inline source/renderer/app/routes-config.js:34]: `ADD` should be renamed to `REGISTRATION`.
Route should be updated to '/voting/registration',

File: source/renderer/app/config/votingConfig.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+// @flow
+export const MIN_VOTING_FUNDS = 7950; // 7950 ADA | unit: ADA
+export const VOTING_FEE_FOR_CALCULATE = 1; // 1 ADA | unit: ADA
+export const PIN_CODE_LENGTH = 4;
```
Comment (nikolaglumac) [inline source/renderer/app/config/votingConfig.js:4]: `PIN_CODE_LENGTH ` should be renamed to `VOTING_PIN_CODE_LENGTH`

File: source/renderer/app/config/votingConfig.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+// @flow
+export const MIN_VOTING_FUNDS = 7950; // 7950 ADA | unit: ADA
+export const VOTING_FEE_FOR_CALCULATE = 1; // 1 ADA | unit: ADA
```
Comment (nikolaglumac) [inline source/renderer/app/config/votingConfig.js:3]: `VOTING_FEE_FOR_CALCULATE` should be renamed to `VOTING_FEE_CALCULATION_AMOUNT`

File: source/renderer/app/stores/VotingStore.js [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+  @action setIsVotingRegistrationTransactionApproved = (value: boolean) => {
+    this.isVotingRegistrationTransactionApproved = value;
+  };
+
+  @action setVotingRegistrationKey = (value: any) => {
+    this.votingRegistrationKey = value;
+  };
+
+  @action setQrCode = (value: string | null) => {
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:88]: value type declaration should be `?string`

File: source/renderer/app/stores/VotingStore.js [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+  @action setIsVotingRegistrationTransactionPending = (value: boolean) => {
+    this.isVotingRegistrationTransactionPending = value;
+  };
+
+  @action setIsVotingRegistrationTransactionApproved = (value: boolean) => {
+    this.isVotingRegistrationTransactionApproved = value;
+  };
+
+  @action setVotingRegistrationKey = (value: any) => {
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:84]: value type declaration should be improved.

File: source/renderer/app/stores/VotingStore.js [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+import wallet from '../utils/wallet';
+import { VOTING_COUNTDOWN_INTERVAL } from '../config/votingConfig';
+
+export default class VotingStore extends Store {
+  @observable selectedVotingWalletId: ?string = null;
+  @observable transactionId: ?string = null;
+  @observable isVotingRegistrationTransactionPending: boolean = false;
+  @observable isVotingRegistrationTransactionApproved: boolean = false;
+  @observable votingRegistrationKey: any = null;
+  @observable qrCode: string | null = null;
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:19]: Type declaration should be improved.

File: source/renderer/app/stores/VotingStore.js [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+import { formattedArrayBufferToHexString } from '../utils/formatters';
+import wallet from '../utils/wallet';
+import { VOTING_COUNTDOWN_INTERVAL } from '../config/votingConfig';
+
+export default class VotingStore extends Store {
+  @observable selectedVotingWalletId: ?string = null;
+  @observable transactionId: ?string = null;
+  @observable isVotingRegistrationTransactionPending: boolean = false;
+  @observable isVotingRegistrationTransactionApproved: boolean = false;
+  @observable votingRegistrationKey: any = null;
```
Comment (nikolaglumac) [inline source/renderer/app/stores/VotingStore.js:18]: Type declaration should be improved.

File: source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+    return (
+      <div className={className}>
+        <div className={contentClassName}>
+          <p className={styles.description}>
+            <FormattedHTMLMessage {...messages.description} />
+          </p>
+          <WalletsDropdown
+            className={walletSelectClasses}
+            label={intl.formatMessage(messages.selectWalletInputLabel)}
+            numberOfStakePools={numberOfStakePools}
```
Comment (nikolaglumac) [inline source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js:158]: This should be removed as it is not needed.

File: source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+  };
+
+  render() {
+    const { intl } = this.context;
+    const { selectedWalletId } = this.state;
+    const {
+      wallets,
+      minVotingFunds,
+      isWalletAcceptable,
+      numberOfStakePools,
```
Comment (nikolaglumac) [inline source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js:108]: This should be removed as it is not needed.

File: source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+  continueButtonLabel: {
+    id: 'voting.votingAdd.chooseWallet.step.continueButtonLabel',
+    defaultMessage: '!!!Continue',
+    description:
+      'Label for continue button on the voting add "choose wallet" step.',
+  },
+});
+
+type Props = {
+  numberOfStakePools: number,
```
Comment (nikolaglumac) [inline source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js:61]: This should be removed as it is not needed.

File: source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+    const { amount, reward, isRestoring } = selectedWallet || {};
+
+    let errorMessage;
+    if (selectedWallet && !isWalletAcceptable(amount, reward)) {
+      // Wallet is restoring
+      if (isRestoring) errorMessage = messages.errorRestoringWallet;
+      // Wallet only has Reward balance
+      else if (!amount.isZero() && amount.equals(reward))
+        errorMessage = messages.errorMinVotingFundsRewardsOnly;
+      // Wallet balance < min delegation funds
```
Comment (nikolaglumac) [inline source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsChooseWallet.js:127]: Comment should be updated.

File: source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsQrCode.js [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    id: 'voting.votingAdd.qrCode.step.warningMessage',
+    defaultMessage:
+      '!!!If you lose access to the QR code, you will need to register again in order to generate a new QR code.',
+    description:
+      'Qr code warning messages of use on the voting add "qr code" step.',
+  },
+});
+
+type Props = {
+  qrCode: string | null,
```
Comment (nikolaglumac) [inline source/renderer/app/components/voting/voting-add-wizard/VotingAddStepsQrCode.js:46]: Type declaration should be improved.

**Review by nikolaglumac — COMMENTED** [review]
Alongside the "todo" section in the PR's description, we need to resolve all the following issues:

**Review by ManusMcCole — APPROVED** [review]

**Review by a-rukin — APPROVED** [review]
"A wizard is never late, nor is he early, he arrives precisely when he means to."
🧙

**Review by gnpf — APPROVED** [review]

**Pull request merged**